### PR TITLE
Add ClearML tracker

### PR DESCRIPTION
**Pull request by eugen-ajechiloae-clearml**
Adds ClearML to the supported trackers.

### CODE REVIEW

**Comment by eugen-ajechiloae-clearml** [comment]
Hi @muellerzr ! Managed to fix the styling issues. Thanks for the instructions!

File: src/accelerate/tracking.py [hunk 1]
```diff
@@ -681,12 +685,187 @@ def finish(self):
...
+            return
+
+        if "CLEARML_PROJECT" in os.environ:
+            task_init_kwargs.setdefault("project_name", os.environ["CLEARML_PROJECT"])
+        else:
+            task_init_kwargs.setdefault("project_name", run_name)
+        if "CLEARML_TASK" in os.environ:
+            task_init_kwargs.setdefault("task_name", os.environ["CLEARML_TASK"])
+        else:
+            task_init_kwargs.setdefault("task_name", run_name)
```
Comment (muellerzr) [inline src/accelerate/tracking.py:726]: Let's simplify this:

```suggestion
        task_init_kwargs["project_name"] = os.environ.get("CLEARML_PROJECT", run_name)
        task_init_kwargs["task_name"] = os.environ.get("CLEARML_TASK", run_name)
```

File: src/accelerate/tracking.py [hunk 2]
```diff
@@ -681,12 +685,187 @@ def finish(self):
...
+      `project_name` in `task_init_kwargs`.
+    - **CLEARML_TASK** (`str`, *optional*) - The default ClearML task name. Can be overwritten by setting `task_name`
+      in `task_init_kwargs`.
+
+    Args:
+        run_name (`str`, *optional*):
+            Name of the experiment. If ClearML's `Task.init`'s `task_name` and `project_name` are not specified in
+            kwargs, they will default to this value
+        task_init_kwargs:
+            Kwargs passed along to the `Run.__init__` method.
```
Comment (muellerzr) [inline src/accelerate/tracking.py:703]:
```suggestion
        kwargs:
            Kwargs passed along to the `Run.__init__` method.
```

Let's be consistent here and stay with `kwargs`

File: src/accelerate/tracking.py [hunk 3]
```diff
@@ -681,12 +685,187 @@ def finish(self):
...
+    Environment:
+    - **CLEARML_PROJECT** (`str`, *optional*) - The default ClearML project name. Can be overwritten by setting
+      `project_name` in `task_init_kwargs`.
+    - **CLEARML_TASK** (`str`, *optional*) - The default ClearML task name. Can be overwritten by setting `task_name`
+      in `task_init_kwargs`.
+
+    Args:
+        run_name (`str`, *optional*):
+            Name of the experiment. If ClearML's `Task.init`'s `task_name` and `project_name` are not specified in
+            kwargs, they will default to this value
```
Comment (muellerzr) [inline src/accelerate/tracking.py:701]:
```suggestion
        run_name (`str`, *optional*):
            Name of the experiment. By default will try and pull from either `CLEARML_PROJECT` or `CLEARML_TASK` from the environment if set.
```

File: src/accelerate/tracking.py [hunk 4]
```diff
@@ -681,12 +685,187 @@ def finish(self):
...
+    @on_main_process
+    def store_init_configuration(self, values: dict):
+        """
+        Connect configuration dictionary to the Task object. Should be run at the beginning of your experiment.
+
+        Args:
+            values (`dict`):
+                Values to be stored as initial hyperparameters as key-value pairs.
+        """
+        if not self.task:
```
Comment (muellerzr) [inline src/accelerate/tracking.py:742]: Can you explain why this would ever be called?

File: src/accelerate/tracking.py [hunk 5]
```diff
@@ -681,12 +685,187 @@ def finish(self):
...
+                `clearml.Logger.report_scalar` methods.
+        """
+        if not self.task:
+            return
+
+        clearml_logger = self.task.get_logger()
+        for k, v in values.items():
+            if not isinstance(v, (int, float)):
+                logger.warning(
+                    "Trainer is attempting to log a value of "
```
Comment (muellerzr) [inline src/accelerate/tracking.py:771]: We are not using the trainer :)

File: src/accelerate/tracking.py [hunk 6]
```diff
@@ -681,12 +685,187 @@ def finish(self):
...
+            step (`int`, *optional*):
+                The run step. If included, the log will be affiliated with this step.
+            kwargs:
+                Additional key word arguments passed along to the `clearml.Logger.report_table` method.
+        """
+        if not self.task:
+            return
+        if dataframe is None:
+            try:
+                import pandas as pd
```
Comment (muellerzr) [inline src/accelerate/tracking.py:834]: We need a check here for dependencies on `pandas`, as it doesn't come pre-installed with Accelerate. Add a util to check in `imports.py` please

**Review by muellerzr — CHANGES_REQUESTED** [review]
I've done a first pass through this, I'll do another in the morning but it's a good place to start. Also if possible please add some tests to the tracker testing suite for `clearml`

File: src/accelerate/tracking.py (hunk 3 above)
Comment (muellerzr) [reply src/accelerate/tracking.py:701]: In general let's try to prioritize the explicit config first, then fall back on the environmental variables that a user would directly know.

File: src/accelerate/tracking.py (hunk 1 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py (hunk 2 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py (hunk 3 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py (hunk 4 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: Paranoia got the best of me :). Indeed, this should never be called. I removed this section

File: src/accelerate/tracking.py (hunk 5 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: changed to Accelerator

File: src/accelerate/tracking.py (hunk 6 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py:828]: done

File: src/accelerate/tracking.py [hunk 7]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+            title, series = ClearMLTracker._get_title_series(k)
+            clearml_logger.report_scalar(title=title, series=series, value=v, iteration=step, **kwargs)
+
+    @on_main_process
+    def log_images(self, values: dict, step: Optional[int] = None, **kwargs):
+        """
+        Logs `images` to the current run.
+
+        Args:
+            values (Dictionary `str` to `List` of `np.ndarray` or `PIL.Image`):
```
Comment (muellerzr) [inline src/accelerate/tracking.py:782]:
```suggestion
            values (`Dict[str, List[Union[np.ndarray, PIL.Image]]`):
```

File: src/accelerate/tracking.py [hunk 8]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+        step: Optional[int] = None,
+        **kwargs,
+    ):
+        """
+        Log a Table to the task. Can be defined eitherwith `columns` and `data` or with `dataframe`.
+
+        Args:
+            table_name (`str`):
+                The name of the table
+            columns (List of `str`'s *optional*):
```
Comment (muellerzr) [inline src/accelerate/tracking.py:810]:
```suggestion
            columns (list of `str`, *optional*):
```

File: src/accelerate/tracking.py [hunk 9]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+                The name of the columns on the table
+            data (List of List of Any data type *optional*):
```
Comment (muellerzr) [inline src/accelerate/tracking.py:812]:
```suggestion
            data (List of List of Any data type, *optional*):
```

File: src/accelerate/tracking.py [hunk 10]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+                The data to be logged in the table
+            dataframe (Any data type *optional*):
```
Comment (muellerzr) [inline src/accelerate/tracking.py:814]:
```suggestion
            dataframe (Any data type, *optional*):
```

File: src/accelerate/tracking.py [hunk 11]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+            step (`int`, *optional*):
+                The run step. If included, the log will be affiliated with this step.
+            kwargs:
+                Additional key word arguments passed along to the `clearml.Logger.report_table` method.
+        """
+        if not is_pandas_available():
+            if not self._pandas_warning_sent:
+                logger.warning("ClearMLTracker.log_table requires pandas to be installed")
+                self._pandas_warning_sent = True
+            return
```
Comment (muellerzr) [inline src/accelerate/tracking.py:825]:
```suggestion
        if not is_pandas_available():
            raise ImportError("`ClearMLTracker.log_table`" requires `pandas` be installed. Please fix this by running `pip install pandas`)
```

Explicit block with solution, not warning and failure.

File: src/accelerate/tracking.py [hunk 12]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+        if dataframe is None:
+            try:
+                import pandas as pd
+
+                if columns is None or data is None:
+                    logger.warning(
+                        "ClearMLTracker.log_table requires that columns and data to be supplied if dataframe is None"
+                    )
+                    return
```
Comment (muellerzr) [inline src/accelerate/tracking.py:834]: Do not warn, raise error instead.

```suggestion
                    raise ValueError("`ClearMLTracker.log_table` requires that `columns` and `data` to be supplied if `dataframe` is `None`")
```

**Review by muellerzr — COMMENTED** [review]
Thanks! We're getting closer!

File: src/accelerate/tracking.py [hunk 13]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+
+        current_task = Task.current_task()
+        self._initialized_externally = False
+        if current_task:
+            self._initialized_externally = True
+            self.task = current_task
+            return
+
+        kwargs.setdefault("project_name", os.environ.get("CLEARML_PROJECT", run_name))
+        kwargs.setdefault("task_name", os.environ.get("CLEARML_TASK", run_name))
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py:716]: I would like people to also be able to set the `project_name` via `kwargs` directly. The priority order I had in mind is:
explicit via `kwargs` -> env vars -> run name (default).
Doing something like:
```py
kwargs.update({"project_name": os.environ.get("CLEARML_PROJECT", run_name), "task_name", os.environ.get("CLEARML_TASK", run_name)}
```
would not allow that

File: src/accelerate/tracking.py [hunk 14]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+                Additional key word arguments passed along to the `clearml.Logger.report_table` method.
+        """
+        if not is_pandas_available():
+            if not self._pandas_warning_sent:
+                logger.warning("ClearMLTracker.log_table requires pandas to be installed")
+                self._pandas_warning_sent = True
+            return
+        if dataframe is None:
+            try:
+                import pandas as pd
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py [hunk 15]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+
+                if columns is None or data is None:
+                    logger.warning(
+                        "ClearMLTracker.log_table requires that columns and data to be supplied if dataframe is None"
+                    )
+                    return
+                dataframe = pd.DataFrame({column: data_entry for column, data_entry in zip(columns, data)})
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: That also works, thank you!

File: src/accelerate/tracking.py [hunk 16]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: tests/test_tracking.py [hunk 17]
```diff
@@ -250,6 +252,118 @@ def test_log(self):
         self.assertEqual(self.get_value_from_key(list_of_json, "my_text"), "some_value")
 
 
+@require_clearml
+class ClearMLTest(unittest.TestCase):
+    @staticmethod
+    def _get_offline_dir(accelerator):
+        from clearml.config import get_offline_dir
+
+        return get_offline_dir(task_id=accelerator.get_tracker("clearml", unwrap=True).id)
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply tests/test_tracking.py:267]: Done

File: src/accelerate/tracking.py [hunk 18]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+    Environment:
+    - **CLEARML_PROJECT** (`str`, *optional*) - The default ClearML project name. Can be overwritten by setting
+      `project_name` in `task_init_kwargs`.
+    - **CLEARML_TASK** (`str`, *optional*) - The default ClearML task name. Can be overwritten by setting `task_name`
+      in `task_init_kwargs`.
+
+    Args:
+        run_name (`str`, *optional*):
+            Name of the experiment. By default will try and pull from either `CLEARML_PROJECT` or `CLEARML_TASK` from
+            the environment if set.
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py [hunk 19]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
 
+class ClearMLTracker(GeneralTracker):
+    """
+    A `Tracker` class that supports `clearml`. Should be initialized at the start of your script.
+
+    Environment:
+    - **CLEARML_PROJECT** (`str`, *optional*) - The default ClearML project name. Can be overwritten by setting
+      `project_name` in `task_init_kwargs`.
+    - **CLEARML_TASK** (`str`, *optional*) - The default ClearML task name. Can be overwritten by setting `task_name`
+      in `task_init_kwargs`.
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py [hunk 20]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+        ints or floats
+
+        Args:
+            values (`dict`):
+                Values to be logged as key-value pairs. If the key starts with 'eval_'/'test_'/'train_',the value will
+                be reported under the 'eval'/'test'/'train' series and the respective prefix will be removed.
+                Otherwise, the value will be reported under the 'train' series, and no prefix will be removed.
+            step (`int`, *optional*):
+                If None (default), the values will be reported as single values. If specified, the values will be
+                reported as scalars, with the iteration number equal to `step`.
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py (hunk 8 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py (hunk 7 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: I did this. I could also add this to the signature if you want me to. We could import PIL this way:
```py
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    import PIL
```

File: src/accelerate/tracking.py (hunk 9 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py (hunk 10 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py [hunk 21]
```diff
@@ -681,12 +686,183 @@ def finish(self):
...
+    def log(self, values: dict, step: Optional[int] = None, **kwargs):
+        """
+        Logs `values` dictionary to the current run. The dictionary keys must be strings. The dictionary values must be
+        ints or floats
+
+        Args:
+            values (`dict`):
+                Values to be logged as key-value pairs. If the key starts with 'eval_'/'test_'/'train_',the value will
+                be reported under the 'eval'/'test'/'train' series and the respective prefix will be removed.
+                Otherwise, the value will be reported under the 'train' series, and no prefix will be removed.
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (muellerzr) [reply src/accelerate/tracking.py:716]: Got it, okay. We can keep `setdefault` then.

File: src/accelerate/tracking.py [hunk 22]
```diff
@@ -681,12 +686,171 @@ def finish(self):
...
+class ClearMLTracker(GeneralTracker):
+    """
+    A `Tracker` class that supports `clearml`. Should be initialized at the start of your script.
+
+    Args:
+        run_name (`str`, *optional*):
+            Name of the experiment. Environment variables `CLEARML_PROJECT` and `CLEARML_TASK` have priority over this
+            argument.
+        kwargs:
+            Kwargs passed along to the `Run.__init__` method.
```
Comment (muellerzr) [inline src/accelerate/tracking.py:698]:
```suggestion
            Kwargs passed along to the `Task.__init__` method.
```

**Review by muellerzr — APPROVED** [review]
Thanks! This is looking great, thanks for iterating with me over it. cc @BenjaminBossan

File: src/accelerate/tracking.py (hunk 22 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py [hunk 23]
```diff
@@ -681,12 +686,171 @@ def finish(self):
...
+        if current_task:
+            self._initialized_externally = True
+            self.task = current_task
+            return
+
+        kwargs.setdefault("project_name", os.environ.get("CLEARML_PROJECT", run_name))
+        kwargs.setdefault("task_name", os.environ.get("CLEARML_TASK", run_name))
+        self.task = Task.init(**kwargs)
+
+        self._pandas_warning_sent = False
```
Comment (BenjaminBossan) [inline src/accelerate/tracking.py:719]: What is the purpose of this attribute?

File: src/accelerate/tracking.py [hunk 24]
```diff
@@ -681,12 +686,171 @@ def finish(self):
...
+        ints or floats
+
+        Args:
+            values (`Dict[str, Union[int, float]]`):
+                Values to be logged as key-value pairs. If the key starts with 'eval_'/'test_'/'train_', the value will
+                be reported under the 'eval'/'test'/'train' series and the respective prefix will be removed.
+                Otherwise, the value will be reported under the 'train' series, and no prefix will be removed.
+            step (`int`, *optional*):
+                If specified, the values will be reported as scalars, with the iteration number equal to `step`.
+                Otherwise they will be reported as single values.
```
Comment (BenjaminBossan) [inline src/accelerate/tracking.py:749]: To me, the distinction is not clear. But if it's generally known to ClearML users, I'm fine with not explaining this further.

File: src/accelerate/tracking.py [hunk 25]
```diff
@@ -681,12 +686,171 @@ def finish(self):
...
+                The data to be logged in the table
+            step (`int`, *optional*):
+                The run step. If included, the log will be affiliated with this step.
+            kwargs:
+                Additional key word arguments passed along to the `clearml.Logger.report_table` method.
+        """
+        if dataframe is None:
+            if not is_pandas_available():
+                raise ImportError(
+                    "`ClearMLTracker.log_table` requires `pandas` be installed. Please fix this by running `pip install pandas`"
```
Comment (BenjaminBossan) [inline src/accelerate/tracking.py:818]:
```suggestion
                    "`ClearMLTracker.log_table` requires `pandas` to be installed. Please fix this by running `pip install pandas`"
```

File: src/accelerate/tracking.py [hunk 26]
```diff
@@ -681,12 +686,171 @@ def finish(self):
...
+                )
+            import pandas as pd
+
+            if columns is None or data is None:
+                raise ValueError(
+                    "`ClearMLTracker.log_table` requires that `columns` and `data` to be supplied if `dataframe` is `None`"
+                )
+            dataframe = pd.DataFrame(data, columns=columns)
+        title, series = ClearMLTracker._get_title_series(table_name)
+        self.task.get_logger().report_table(title=title, series=series, table_plot=dataframe, iteration=step, **kwargs)
```
Comment (BenjaminBossan) [inline src/accelerate/tracking.py:828]: ClearML [docs](https://clear.ml/docs/latest/docs/references/sdk/logger#report_table) say that:

> Pandas DataFrame or Table as list of rows (list)

are supported. Not sure what they mean by "Table", but if it's just a list of rows, can we get by without pandas here?

File: tests/test_tracking.py [hunk 27]
```diff
@@ -250,6 +253,127 @@ def test_log(self):
...
+        from clearml import Task
+
+        Task.set_offline(True)
+        accelerator = Accelerator(log_with="clearml")
+        accelerator.init_trackers("test_project_with_log_images")
+
+        base_image = np.eye(256, 256, dtype=np.uint8) * 255
+        images = {
+            "train_image": base_image,
+            "test_image": np.concatenate((np.atleast_3d(base_image), np.zeros((256, 256, 2), dtype=np.uint8)), axis=2),
```
Comment (BenjaminBossan) [inline tests/test_tracking.py:335]: Could this be assigned to `base_image_3d` or something like this, that would be more readable.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks a lot, in general, this looks good. Failing tests seem to be unrelated, but we should re-run CI once that issue has been resolved.

I noticed that we need an entry in `docs/source/package_reference/tracking.md` for docs to show the new tracker. Also, the list in `docs/source/usage_guides/tracking.md` needs to be updated (please also add `AimTracker`).

Moreover, I saw that other trackers have a lot more debug logs, should they be added here too for consistency?

**Comment by BenjaminBossan** [comment]
@eugen-ajechiloae-clearml Merging with/rebasing on main should fix the failing CI.

File: src/accelerate/tracking.py (hunk 23 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: removed

File: src/accelerate/tracking.py (hunk 24 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py:746]: I think that ClearML users should be aware of the distinction

File: src/accelerate/tracking.py (hunk 25 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: done

File: src/accelerate/tracking.py (hunk 26 above)
Comment (eugen-ajechiloae-clearml) [reply src/accelerate/tracking.py]: You are correct :+1: . I made some changes to remove pandas as a req.

File: tests/test_tracking.py (hunk 27 above)
Comment (eugen-ajechiloae-clearml) [reply tests/test_tracking.py]: done

**Comment by BenjaminBossan** [comment]
@eugen-ajechiloae-clearml LMK once the PR is ready for the next review.

**Comment by eugen-ajechiloae-clearml** [comment]
@BenjaminBossan the PR is ready for the next review now

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for updating the PR. Nice that this now works without pandas. It's almost ready, I only have a small comment, and also the doc entries that I mentioned in my previous review are still missing, right?

**Comment by BenjaminBossan** [comment]
@eugen-ajechiloae-clearml could you please address this:

> I noticed that we need an entry in docs/source/package_reference/tracking.md for docs to show the new tracker. Also, the list in docs/source/usage_guides/tracking.md needs to be updated (please also add AimTracker).

Apart from that, we're ready to merge.

**Comment by eugen-ajechiloae-clearml** [comment]
Hi @BenjaminBossan ! I updated the docs as well

**Review by BenjaminBossan — APPROVED** [review]
Thanks for adding the tracker to the docs. LGTM.

**Pull request merged**